<a href="https://colab.research.google.com/github/Bharat745/H2O/blob/master/Random_Forest_iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Random Forest Iris dataset

## Installation and importing necessary libraries and the dataset


In [1]:
# Installing java version for running H2O on colab
! apt-get install default-jre
!java -version

Reading package lists... Done
Building dependency tree       
Reading state information... Done
default-jre is already the newest version (2:1.11-68ubuntu1~18.04.1).
default-jre set to manually installed.
The following package was automatically installed and is no longer required:
  libnvidia-common-410
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 16 not upgraded.
openjdk version "11.0.3" 2019-04-16
OpenJDK Runtime Environment (build 11.0.3+7-Ubuntu-1ubuntu218.04.1)
OpenJDK 64-Bit Server VM (build 11.0.3+7-Ubuntu-1ubuntu218.04.1, mixed mode, sharing)


In [2]:
# Start and connect to local H2O cluster
! pip install h2o
import h2o
h2o.init(nthreads = -1)

     |████████████████████████████████| 122.4MB 171kB/s 
  Stored in directory: /root/.cache/pip/wheels/fe/31/5a/d0a96f4ab19a49d5381707eb3017b1a41ec89acbeff54a532c
Successfully built h2o
Checking whether there is an H2O instance running at http://localhost:54321 ..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "11.0.3" 2019-04-16; OpenJDK Runtime Environment (build 11.0.3+7-Ubuntu-1ubuntu218.04.1); OpenJDK 64-Bit Server VM (build 11.0.3+7-Ubuntu-1ubuntu218.04.1, mixed mode, sharing)
  Starting server from /usr/local/lib/python3.6/dist-packages/h2o/backend/bin/h2o.jar
  Ice root: /tmp/tmpwk1ygafy
  JVM stdout: /tmp/tmpwk1ygafy/h2o_unknownUser_started_from_python.out
  JVM stderr: /tmp/tmpwk1ygafy/h2o_unknownUser_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         02 secs
H2O cluster timezone:       Etc/UTC
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.5
H2O cluster version age:    16 days
H2O cluster name:           H2O_from_python_unknownUser_243a22
H2O cluster total nodes:    1
H2O cluster free memory:    3 Gb
H2O cluster total cores:    2
H2O cluster allowed cores:  2
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.8 final
--------------------------  ---------------------------------------------------

In [3]:
#Import dataset
url = "http://h2o-public-test-data.s3.amazonaws.com/smalldata/iris/iris_wheader.csv"
iris= h2o.import_file(url)

Parse progress: |█████████████████████████████████████████████████████████| 100%


## Data preparation and descriptive statistics

In [0]:
# Splitting the dataset to train and test
train, test = iris.split_frame([0.8])

In [5]:
#looking at the datasets
train.summary()
test.summary()

sepal_len sepal_wid petal_len petal_wid class 
 
 
 type real real real real enum 
 mins 4.3 2.0 1.0 0.1 
 mean 5.766129032258063 3.025806451612904 3.675 1.1685483870967746 
 maxs 7.7 4.1 6.9 2.5 
 sigma 0.8046498098854236 0.4271042537077735 1.760341107004018 0.768312154495388 
 zeros 0 0 0 0 
 missing 0 0 0 0 0 
 0 5.1 3.5 1.4 0.2 Iris-setosa 
 1 4.9 3.0 1.4 0.2 Iris-setosa 
 2 4.6 3.1 1.5 0.2 Iris-setosa 
 3 5.0 3.6 1.4 0.2 Iris-setosa 
 4 5.4 3.9 1.7 0.4 Iris-setosa 
 5 4.6 3.4 1.4 0.3 Iris-setosa 
 6 5.0 3.4 1.5 0.2 Iris-setosa 
 7 4.4 2.9 1.4 0.2 Iris-setosa 
 8 4.9 3.1 1.5 0.1 Iris-setosa 
 9 5.4 3.7 1.5 0.2 Iris-setosa

sepal_len sepal_wid petal_len petal_wid class 
 
 
 type real real real real enum 
 mins 4.7 2.5 1.3 0.1 
 mean 6.211538461538463 3.1884615384615382 4.157692307692307 1.3423076923076922 
 maxs 7.9 4.4 6.7 2.3 
 sigma 0.8547874229663994 0.4475059088565631 1.7629913390423853 0.7354851878443655 
 zeros 0 0 0 0 
 missing 0 0 0 0 0 
 0 4.7 3.2 1.3 0.2 Iris-setosa 
 1 5.7 4.4 1.5 0.4 Iris-setosa 
 2 5.0 3.4 1.6 0.4 Iris-setosa 
 3 5.2 3.5 1.5 0.2 Iris-setosa 
 4 5.5 4.2 1.4 0.2 Iris-setosa 
 5 4.9 3.1 1.5 0.1 Iris-setosa 
 6 5.1 3.8 1.9 0.4 Iris-setosa 
 7 6.4 3.2 4.5 1.5 Iris-versicolor 
 8 6.9 3.1 4.9 1.5 Iris-versicolor 
 9 6.5 2.8 4.6 1.5 Iris-versicolor

In [6]:
train.nrows

124

In [7]:
test.nrows

26

## Analytics and Modelling

In [0]:
# Importing Random forest estimator
from h2o.estimators.random_forest import H2ORandomForestEstimator

In [0]:
# Define the features and target
features = list(test.columns)
target = "class"

In [10]:
# Train the Random Forest Model
mRF = H2ORandomForestEstimator(seed= 1234)
mRF.train(x= features, y = target ,training_frame = train)
print(mRF)

drf Model Build progress: |███████████████████████████████████████████████| 100%
Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  DRF_model_python_1562340797463_1


ModelMetricsMultinomial: drf
** Reported on train data. **

MSE: 0.03145618375278179
RMSE: 0.1773589122451471
LogLoss: 0.10010960350530813
Mean Per-Class Error: 0.04939024390243902
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error      Rate
-------------  -----------------  ----------------  ---------  -------
43             0                  0                 0          0 / 43
0              38                 3                 0.0731707  3 / 41
0              3                  37                0.075      3 / 40
43             41                 40                0.0483871  6 / 124

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.951613
2    1
3    1

Scoring History: 


timestamp            duration    number_of_trees    training_rmse        training_logloss     training_classification_error
---  -------------------  ----------  -----------------  -------------------  -------------------  -------------------------------
     2019-07-05 15:34:12  0.032 sec   0.0                nan                  nan                  nan
     2019-07-05 15:34:12  0.213 sec   1.0                0.2672612419124244   2.4670554567793346   0.07142857142857142
     2019-07-05 15:34:12  0.233 sec   2.0                0.24691458232921557  1.8174917538814388   0.07792207792207792
     2019-07-05 15:34:12  0.250 sec   3.0                0.21559139121216433  1.3856125252363445   0.0594059405940594
     2019-07-05 15:34:12  0.268 sec   4.0                0.20622085324877049  1.2420560190883092   0.05309734513274336
---  ---                  ---         ---                ---                  ---                  ---
     2019-07-05 15:34:12  0.750 sec   46.0               0.1771788214622026   0.09928482872456164  0.04838709677419355
     2019-07-05 15:34:12  0.757 sec   47.0               0.1779515846964051   0.10046486929326098  0.04838709677419355
     2019-07-05 15:34:13  0.763 sec   48.0               0.1768701592604855   0.09928657055612242  0.04838709677419355
     2019-07-05 15:34:13  0.777 sec   49.0               0.1783338514754394   0.10064351444087312  0.04838709677419355
     2019-07-05 15:34:13  0.785 sec   50.0               0.1773589122451471   0.10010960350530813  0.04838709677419355


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
petal_len   1720.2                 1                    0.482589
petal_wid   1327.92                0.771959             0.372539
sepal_len   405.561                0.235765             0.113777
sepal_wid   110.838                0.0644333            0.0310948

We see from our model above that 6 out of 124 were predicted wrongly, which gives a 95% accuracy on our training data. We see that the model perfectly predicts Iris-sentosa, but wrongly predicts the versicolor and verginica. We also see that model made 50 trees to train. We see from the variable importance data that petal_len is the most important variable while making the predictions followed by others. 

## Results and next steps

In [11]:
# Using the Random Forest model to make predictions
p = mRF.predict(test)

drf prediction progress: |████████████████████████████████████████████████| 100%


In [12]:
p

predict Iris-setosa Iris-versicolor Iris-virginica 
 
 
 Iris-setosa 0.995243 0.00191393 0.00284355 
 Iris-setosa 0.99515 0.00195125 0.00289901 
 Iris-setosa 0.995243 0.00191393 0.00284355 
 Iris-setosa 0.995243 0.00191393 0.00284355 
 Iris-setosa 0.99515 0.00195125 0.00289901 
 Iris-setosa 0.995243 0.00191393 0.00284355 
 Iris-setosa 0.995243 0.00191393 0.00284355 
 Iris-versicolor 0.000774913 0.956605 0.0426202 
 Iris-versicolor 0.000764915 0.747935 0.2513 
 Iris-versicolor 0.000792184 0.97597 0.0232374

Predictions are made with a probablity of all the catagorical variables. The model classifies based on the probablity of the highest catagory.

In [13]:
# Checking the performance of the model
mRF.model_performance(test)


ModelMetricsMultinomial: drf
** Reported on test data. **

MSE: 0.03504376799709709
RMSE: 0.1871998076844554
LogLoss: 0.11477835352694324
Mean Per-Class Error: 0.037037037037037035
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error      Rate
-------------  -----------------  ----------------  ---------  ------
7              0                  0                 0          0 / 7
0              8                  1                 0.111111   1 / 9
0              0                  10                0          0 / 10
7              8                  11                0.0384615  1 / 26

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.961538
2    1
3    1

We see from results that 1 of 26 predictions made was wrong. which gives us a prediction accuracy of 96% on our test data set.

In [0]:
# To know the details of each parameters
help(h2o.estimators.random_forest.H2ORandomForestEstimator)